lib/CPathfinder.cpp

```diff
@@ -455,13 +526,17 @@ CPathsInfo::CPathsInfo( const int3 &Sizes )
 :sizes(Sizes)
 {
 	hero = nullptr;
-	nodes = new CGPathNode**[sizes.x];
+	nodes = new CGPathNode***[sizes.x];
 	for(int i = 0; i < sizes.x; i++)
 	{
-		nodes[i] = new CGPathNode*[sizes.y];
+		nodes[i] = new CGPathNode**[sizes.y];
 		for(int j = 0; j < sizes.y; j++)
 		{
-			nodes[i][j] = new CGPathNode[sizes.z];
+			nodes[i][j] = new CGPathNode*[sizes.z];
+			for (int z = 0; z < sizes.z; z++)
+			{
+				nodes[i][j][z] = new CGPathNode[EPathfindingLayer::NUM_LAYERS];
```

Comment:
I suggest to redo allocation to boost::multi_array (and remove array of NUM_LAYERS along the way). Current code will result in up to 144*144*2 - over 40 000 allocations. This causes:
- increased memory usage since each allocation needs some system data as well
- increased CPU usage - allocations are slow
- reduced CPU cache efficiency - memory blocks are not guaranteed to be continios and can be heavily fragmented. Plus overhead due to system data I mentioned above.

Besides - this is C++, not C. There is rarely need to use C-style dynamic array. std::vector<> is superior in most cases.